File: pythonCode/faceLandmarkDetector.py
import cv2
import dlib
import sys
import numpy as np

def drawPoints(image, landmarks):
    for part in landmarks.parts():
        cv2.circle(image, (part.x, part.y), 3, (0, 255, 255), -1)

#define face detector
faceDetector = dlib.get_frontal_face_detector()

#define landmark detector and load the landmark model
landmarkDetector = dlib.shape_predictor("../dlibAndModel/shape_predictor_68_face_landmarks.dat")

#read images with single and multiple persons
imageSingle = cv2.imread("../assets/anish.jpg")
imageMultiple = cv2.imread("../assets/anish2.jpg")

#create images clone to woek on
imageSingleClone = imageSingle.copy()
imageMultipleClone = imageMultiple.copy()

#convert image to dlibs image format
dlibSingleImage = cv2.cvtColor(imageSingleClone, cv2.COLOR_BGR2RGB)
dlibMultipleImage = cv2.cvtColor(imageMultipleClone, cv2.COLOR_BGR2RGB)

#detect faces in the image
facesSingle = faceDetector(dlibSingleImage, 0)
facesMultiple = faceDetector(dlibMultipleImage, 0)

#loop over all the faces detected
for i in range(0, len(facesSingle)):
    dlibRect = dlib.rectangle(int(facesSingle[i].left()), int(facesSingle[i].top()), int(facesSingle[i].right()), int(facesSingle[i].bottom()))
    landmarks = landmarkDetector(dlibSingleImage, dlibRect)
    drawPoints(imageSingleClone, landmarks)

for i in range(0, len(facesMultiple)):
    dlibRect = dlib.rectangle(int(facesMultiple[i].left()), int(facesMultiple[i].top()), int(facesMultiple[i].right()), int(facesMultiple[i].bottom()))
    landmarks = landmarkDetector(dlibMultipleImage, dlibRect)
    drawPoints(imageMultipleClone, landmarks)

#create windows to display images
cv2.namedWindow("single person image", cv2.WINDOW_NORMAL)
cv2.namedWindow("single person landmarks", cv2.WINDOW_NORMAL)
cv2.namedWindow("multiple person image", cv2.WINDOW_NORMAL)
cv2.namedWindow("multiple person landmarks", cv2.WINDOW_NORMAL)

#display images
cv2.imshow("single person image", imageSingle)
cv2.imshow("single person landmarks", imageSingleClone)
cv2.imshow("multiple person image", imageMultiple)
cv2.imshow("multiple person landmarks", imageMultipleClone)

#press esc to exit the program
cv2.waitKey(0)

#close all the opened windows
cv2.destroyAllWindows()

File: cPlusPlusCode/faceLandmarkDetector.cpp
#include<dlib/image_processing.h>
#include<dlib/image_processing/frontal_face_detector.h>
#include<dlib/opencv.h>
#include<opencv2/imgproc.hpp>
#include<opencv2/highgui.hpp>
#include<iostream>

using namespace std;
using namespace dlib;

//draw points on the image
void drawPoints(cv::Mat &image, full_object_detection landmarks){
    for(int i=0; i<landmarks.num_parts(); i++){
        cv::circle(image, cv::Point(landmarks.part(i).x(), landmarks.part(i).y()), 3, cv::Scalar(0, 255, 255), -1);
    }
}

int main(){
    //get face detector
    frontal_face_detector faceDetector = get_frontal_face_detector();

    //define landmark detector
    shape_predictor landmarkDetector;

    //load the face landmark model
    deserialize("../dlibAndModel/shape_predictor_68_face_landmarks.dat") >> landmarkDetector;

    //Read images with single and mulitple persons
    cv::Mat imageSingle = cv::imread("../assets/anish.jpg");
    cv::Mat imageMultiple = cv::imread("../assets/anish2.jpg");

    //create images clone to work with
    cv::Mat imageSingleClone=imageSingle.clone();
    cv::Mat imageMultipleClone=imageMultiple.clone();

    //convert opencv image format to dlib image format
    cv_image<bgr_pixel> dlibSingleImage(imageSingleClone);
    cv_image<bgr_pixel> dlibMultipleImage(imageMultipleClone);

    //detect faces in the images
    std::vector<rectangle> facesSingle = faceDetector(dlibSingleImage);
    std::vector<rectangle> facesMultiple = faceDetector(dlibMultipleImage);

    //vector to store face landmarks for all the faces
    std::vector<full_object_detection> faceLandmarksSingle;
    std::vector<full_object_detection> faceLandmarksMultiple;

    //Loop over all the faces detected and find face landmarks for all of them
    for(int i=0; i<facesSingle.size(); i++){
        full_object_detection landmarks = landmarkDetector(dlibSingleImage, facesSingle[i]);
        faceLandmarksSingle.push_back(landmarks);
        drawPoints(imageSingleClone, landmarks);
    }
    for(int i=0; i<facesMultiple.size(); i++){
        full_object_detection landmarks = landmarkDetector(dlibMultipleImage, facesMultiple[i]);
        faceLandmarksMultiple.push_back(landmarks);
        drawPoints(imageMultipleClone, landmarks);
    }

    //create windows to display windows
    cv::namedWindow("single person image", cv::WINDOW_NORMAL);
    cv::namedWindow("single face landmarks", cv::WINDOW_NORMAL);
    cv::namedWindow("multiple person image", cv::WINDOW_NORMAL);
    cv::namedWindow("multiple face landmarks", cv::WINDOW_NORMAL);

    //display images
    cv::imshow("single person image", imageSingle);
    cv::imshow("single face landmarks", imageSingleClone);
    cv::imshow("multiple person image", imageMultiple);
    cv::imshow("multiple face landmarks", imageMultipleClone);

    //press esc to exit the program
    cv::waitKey(0);
    
    //close all the opened windows
    cv::destroyAllWindows();

    return 0;
}